rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+reverting, and in every valid case, it eventually reverts with a unique error
+code for every `panic` invocation, which is recorded in the `"errorCodes"`
+section.
+
+## Integration considerations
+
+Manual reverts with custom codes should still be available to developers, but
+discouraged.
+
+The standard library should henceforth strive to use error types for both
+recoverable and irrecoverable errors. Ideally the standard library should
+never produce a revert code that doesn't have a corresponding documented error
+message.
+
+We should also introduce compiler warnings for manually producing revert codes
+that may conflict with either the reserved code range or the auto-generated panic
+code range.
+
+SDK integration of this feature is open ended, but we should at least aim to
+be able to use error message information to decode revert codes and error types
+that are directly returned.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Using a special intrinsic is the main drawback of this solution for a number of reasons.
+
+It muddles how intrinsics are identified since unlike other compiler intrinsics
+`panic` does not have a `__` prefix. 
+This is necessary since it needs to
+be directly invoked at the error site for the location information to be
+meaningful.
+
+It also pollutes the namespace with yet another reserved word.
+
+And ultimately it moves more complexity to the compiler rather than let the user define it.
+This is a shared problem with the error type annotations.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Despite the drawbacks of moving more special behavior the compiler, it has to
+shoulder this complexity in Sway since we do not have sufficiently powerful
+meta-programming features such that developers may be able to define their own
+language features of this caliber.
+
+Given the benefits of having standardized errors, these trade-offs seem
+acceptable, so long as we allow the solution to be extended in the future to
+be more user defined. The `Error` trait is the main vector of such future
+changes in this case.
+
+Letting the errors be populated directly in the bytecode would be a lot simpler,
+but it is not desirable because it extends the size of the binary for no gain.
+
+Moreover, the question of the location of error strings can be asked. Why
+put them in the ABI specification file? ABI files are soon to benefit from
+a registry that will allow users to easily fetch them, they already contain
+type metadata that can be extended, and they ultimately are our equivalent to
+headers, encompassing all the required information to call a contract and get
+a meaningful answer. Errors are part of making this answer meaningful and it is
+typical for error messages to live in headers.
+
+If we do not specify some kind of error standard, debugging any sort of complex
+Sway error will remain extremely difficult and supporting Sway contracts may be
+a nigh impossible task.
```

Comment:
Having a "nigh" in a sentence is nigh impossible not to interpret as a typo at first :smile:  
```suggestion
an almost impossible task.
```